File: backend-spring/src/test/java/com/codequest/service/LearningContextServiceTest.java
package com.codequest.service;

import com.codequest.domain.entity.UserLearningProfile;
import com.codequest.dto.UpdateLearningContextRequest;
import com.codequest.repository.UserLearningProfileRepository;
import com.codequest.security.SecurityContext;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

class LearningContextServiceTest {

    @Mock
    private UserLearningProfileRepository userLearningProfileRepository;

    private LearningContextService learningContextService;

    @BeforeEach
    void setUp() {
        MockitoAnnotations.openMocks(this);
        learningContextService = new LearningContextService(userLearningProfileRepository);
    }

    @Test
    void testUpdateLearningContextCreatesNewIfNotExists() {
        String userId = "test-user";
        SecurityContext.setUserId(userId);

        when(userLearningProfileRepository.findByUserIdAndLanguageAndTrackAndLevelId(userId, "python", "beginner", 1))
                .thenReturn(Optional.empty());
        when(userLearningProfileRepository.save(any())).thenAnswer(invocation -> invocation.getArgument(0));

        var request = new UpdateLearningContextRequest("python", "beginner", 1, "add(a b)", "missing colon");
        var result = learningContextService.updateLearningContext(request);

        assertNotNull(result);
        assertEquals(1, result.getAttempts());
        assertEquals("add(a b)", result.getFailedCommands());
        assertEquals("missing colon", result.getLastError());
        verify(userLearningProfileRepository).save(any());
    }

    @Test
    void testUpdateLearningContextIncrementsAttempts() {
        String userId = "test-user";
        SecurityContext.setUserId(userId);

        UserLearningProfile existingContext = new UserLearningProfile(userId, "python", "beginner", 1);
        existingContext.setAttempts(2);

        when(userLearningProfileRepository.findByUserIdAndLanguageAndTrackAndLevelId(userId, "python", "beginner", 1))
                .thenReturn(Optional.of(existingContext));
        when(userLearningProfileRepository.save(any())).thenAnswer(invocation -> invocation.getArgument(0));

        var request = new UpdateLearningContextRequest("python", "beginner", 1, "new error", "syntax error");
        var result = learningContextService.updateLearningContext(request);

        assertEquals(3, result.getAttempts());
    }
}

File: backend-spring/src/test/kotlin/com/codequest/service/LearningContextServiceTest.kt
package com.codequest.service

import com.codequest.domain.entity.UserLearningProfile
import com.codequest.dto.UpdateLearningContextRequest
import com.codequest.repository.UserLearningProfileRepository
import com.codequest.security.SecurityContext
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

class LearningContextServiceTest {

    private lateinit var userLearningProfileRepository: UserLearningProfileRepository
    private lateinit var learningContextService: LearningContextService

    @BeforeEach
    fun setUp() {
        userLearningProfileRepository = mock()
        learningContextService = LearningContextService(userLearningProfileRepository)
    }

    @Test
    fun `updateLearningContext should create new context if not exists`() {
        val userId = "test-user"
        SecurityContext.setUserId(userId)

        whenever(
            userLearningProfileRepository.findByUserIdAndLanguageAndTrackAndLevelId(
                userId, "python", "beginner", 1
            )
        ).thenReturn(null)

        whenever(userLearningProfileRepository.save(any())).thenAnswer { it.arguments[0] }

        val request = UpdateLearningContextRequest(
            language = "python",
            track = "beginner",
            levelId = 1,
            failedCommands = "add(a b)",
            lastError = "missing colon"
        )

        val result = learningContextService.updateLearningContext(request)

        assertNotNull(result)
        assertEquals(1, result.attempts)
        assertEquals("add(a b)", result.failedCommands)
        assertEquals("missing colon", result.lastError)
        verify(userLearningProfileRepository).save(any())
    }

    @Test
    fun `updateLearningContext should increment attempts`() {
        val userId = "test-user"
        SecurityContext.setUserId(userId)

        val existingContext = UserLearningProfile(
            userId = userId,
            language = "python",
            track = "beginner",
            levelId = 1,
            attempts = 2
        )

        whenever(
            userLearningProfileRepository.findByUserIdAndLanguageAndTrackAndLevelId(
                userId, "python", "beginner", 1
            )
        ).thenReturn(existingContext)

        whenever(userLearningProfileRepository.save(any())).thenAnswer { it.arguments[0] }

        val request = UpdateLearningContextRequest(
            language = "python",
            track = "beginner",
            levelId = 1,
            failedCommands = "new error",
            lastError = "syntax error"
        )

        val result = learningContextService.updateLearningContext(request)

        assertEquals(3, result.attempts)
    }
}

